File: app/core/Bot.hs
{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
module Bot where

import Card
import State
import System.Random
import Helpers (replaceAtIndex)

-- Battlestate -> nível de dificuldade (dl) -> elemento escolhido
makeChoice :: BattleState -> Int -> Int
makeChoice bstate dl = do
    let seed = playerScore bstate
    let weight = wDeck (playerDeck bstate) dl
    let cpu_hand = take 5 (cpuDeck bstate)
    let w_cpu_hand = wHand cpu_hand weight
    let len = length w_cpu_hand

    w_cpu_hand !! nRandom seed (len-1)


wDeck :: [Card] -> Int -> [Int] -- F N A M T
wDeck deck dl = wDeckRecursive deck dl (replicate 5 $ 1+(6*dl))

wDeckRecursive :: [Card] -> Int -> [Int] -> [Int]
wDeckRecursive [] _ w = w
wDeckRecursive (a:as) dl w =
    case elem of
        -- Para cada elemento do tipo indicado presente no deck do jogador, deve-se aumentar o peso de 2 elementos e didx_to_decreaseinuir de 2
        -- Para elem == F, aumentar A e T (posições 2 e 4) e diminuir N e M (1 e 3)
        "fire" -> wDeckRecursive as dl (updateWeight dl [2,4,1,3] w)
        -- Para elem == N, aumentar F e M (posições 0 e 3) e diminuir A e T (2 e 4)
        "nature" -> wDeckRecursive as dl (updateWeight dl [0,3,2,4] w)
        -- Para elem == A, aumentar N e T (posições 1 e 4) e diminuir F e M (0 e 3)
        "water" -> wDeckRecursive as dl (updateWeight dl [1,4,0,3] w)
        -- Para elem == M, aumentar F e A (posições 0 e 2) e diminuir N e T (1 e 4)
        "metal" -> wDeckRecursive as dl (updateWeight dl [0,2,1,4] w)
        -- Para elem == T, aumentar N e M (posições 1 e 3) e diminuir F e A (0 e 2)
        "earth" -> wDeckRecursive as dl (updateWeight dl [1,3,0,2] w)
        
        _ -> [-7]
    where elem = element a

updateWeight :: Int -> [Int] -> [Int] -> [Int]
updateWeight dl [idx_to_increase1, idx_to_increase2, idx_to_decrease1, idx_to_decrease2] weight = do
    let value_to_increase1 = (weight !! idx_to_increase1) + dl
    let w1 = updateArray weight idx_to_increase1 value_to_increase1
    let value_to_increase2 = (weight !! idx_to_increase2) + dl
    let w2 = updateArray w1 idx_to_increase2 value_to_increase2
    let value_to_decrase1 = (weight !! idx_to_decrease1) + dl
    let w3 = updateArray w2 idx_to_decrease1 value_to_decrase1
    let value_to_decrase2 = (weight !! idx_to_decrease2) + dl
    updateArray w3 idx_to_decrease2 value_to_decrase2


updateArray :: [Int] -> Int -> Int -> [Int]
updateArray arr idx newVal = replaceAtIndex arr newVal idx

-- Gera um número aleatório a partir da seed até o numero max, recebidos como parâmetro
-- Semente -> Limite superior -> Numero aleatório
nRandom :: Int -> Int -> Int
nRandom seed max =  mod (31237 * seed) max

-- A ordem agora é FNAMT
-- O que esta função faz é chamar a função recursiva wHandRecursive
-- recebe uma lista de cartas (a mão do player) -> uma lista de pesos (o peso das cartas de fogo está na posição 0, de natureza em 1 e assim por diante)
-- retorna uma lista de inteiros, onde os índices [0...4] são repetidos, por exemplo, se a mão do computador for: [F1,A3,A1,N1,T2] e os pesos forem: [2,2,3,1,2] o retorno será:
-- [0,0,1,1,1,2,2,2,3,3,4,4] -> os dois zeros no início fazem referência à carta na posição 0 (F1) que é de fogo, e fogo tem peso 2, o mesmo vale para os outros elementos
wHand :: [Card] -> [Int] -> [Int]
wHand c w = wHandRecursive 0 c w

wHandRecursive :: Int -> [Card] -> [Int] -> [Int]
wHandRecursive 4 [] [] = []
wHandRecursive 4 _ _ = []
wHandRecursive i cards weights =
    case elem of
        "fire" -> replicate (weights !! 0) i ++ wHandRecursive (i + 1) cards weights
        "nature" -> replicate (weights !! 1) i ++ wHandRecursive (i + 1) cards weights
        "water" -> replicate (weights !! 2) i ++ wHandRecursive (i + 1) cards weights
        "metal" -> replicate (weights !! 3) i ++ wHandRecursive (i + 1) cards weights
        "earth" -> replicate (weights !! 4) i ++ wHandRecursive (i + 1) cards weights
        _ -> [-7]
    where elem = element (cards !! i)

-- Função para fazer o shuffle de um array 
shuffle :: [a] -> IO [a]
shuffle [] = return []
shuffle xs = do randomPosition <- getStdRandom (randomR (0, length xs - 1))
                let (left, (a:right)) = splitAt randomPosition xs
                fmap (a:) (shuffle (left ++ right))

File: app/core/Bot.pl
:- consult('../util/StateManager.pl').

/* 
Pegar o deck do jogador:
get_player_state(D),
nth0(2,D,Deck).
*/

/*
Pegar a mao do bot:
get_bot_state(D)
nth0(2,D,Deck)
*/

indexElem('fire', 0).
indexElem('nature', 1).
indexElem('water', 2).
indexElem('metal', 3).
indexElem('earth', 4).

arrayElem('fire', [2,4,1,3]).
arrayElem('nature', [0,3,2,4]).
arrayElem('water', [1,4,0,3]).
arrayElem('metal', [0,2,1,4]).
arrayElem('earth', [1,3,0,2]).

/* 
Função que seleciona o index de uma das cartas da mão do bot
*/
makeChoice(Level,Id):-
    get_bot_state(D),
    nth0(2,D,Deck),
    slice(0, 5, Deck, Hand),
    get_player_state(PlayerState),
    wDeck(Level,PlayerState,Weight),
    wHand(Hand,Weight, W_cpu_hand),
    length(W_cpu_hand, L),
    random(0,L,Choice),
    nth0(Choice,W_cpu_hand,Id).

/*
Retorna os pesos de cada elemento na ordem FNAMT
*/
wDeck(Level,PlayerState,Weight):-
    nth0(2,PlayerState,Deck),
    V is 1+(6*Level),
    createList(V,5,CurrentWeight),
    wDeckRecursive(Deck, Level, CurrentWeight, Weight).

/*
Varre todo o deck, modificando os pesos de acordo com as cartas encontradas
*/
wDeckRecursive([],_,CurrentWeight,CurrentWeight):-!.
wDeckRecursive([card(id(_), elem(E), power(_))|T], Level, CurrentWeight, Weight):-
    arrayElem(E,Arr),
    updateWeight(I, Arr, CurrentWeight,NCW),
    wDeckRecursive(T,I,NCW,Weight),
    !.

/*
Atualiza a lista de pesos
*/
updateWeight(Level, [Idx_to_increase1, Idx_to_increase2, Idx_to_decrease1, Idx_to_decrease2],CurrentWeight,NewWeight):-
    auxUpdateWeight(Level,Idx_to_increase1,CurrentWeight,New1),
    auxUpdateWeight(Level,Idx_to_increase2,New1,New2),
    NegativeI is Level * (-1),
    auxUpdateWeight(NegativeI,Idx_to_decrease1,New2,New3),
    auxUpdateWeight(NegativeI,Idx_to_decrease2,New3,NewWeight).

auxUpdateWeight(Level,Idx,Current,New):-
    nth0(Idx,Current,Increase),
    Value_to_increase is Increase + I,
    updateArray(Current, Idx, Value_to_increase, New).
    
/*
Atualiza o valor de uma lista na posição Idx por um Novo valor (NV)
*/
updateArray([_|T],0, NV, [NV|T]):-!.
updateArray([H|T],Idx, NV, [H|Return]):-
    NIdx is Idx - 1,
    updateArray(T,NIdx, NV, Return).

/*
Cria uma lista com um valor V repetido N vezes
*/
createList(_, 0, []). % Caso base: quando N é 0, a lista está vazia.
createList(V, N, [V|Resto]) :-
    N > 0, % Garante que N seja maior que 0 para continuar a recursão.
    N1 is N - 1, % Decrementa N para a próxima chamada recursiva.
    createList(V, N1, Resto). % Chama recursivamente para o restante da lista.

/*
Monta uma lista onde cada índice de 0 a 5 [0,5) é repetido um número de vezes igual
ao peso do elemento naquela posição na mão do bot
*/
wHand(Hand,WeightList,WeightHand):-
    wHandRecursive(0,Hand, WeightList, WeightHand).

wHandRecursive(5, _, _, []):-!.
wHandRecursive(I, [card(id(_),elem(E),power(_))|T], WeightList, Return):-
    indexElem(E,Index),
    nth0(Index, WeightList, N),
    createList(I, N, List),
    Ni is I + 1 ,
    wHandRecursive(Ni, T, WeightList, WeightHand),
    append(List,WeightHand,Return).

/*
Função slice auxiliar
*/
slice(0, End, [H|T], [H|Return]):-
    End > 0,
    NewEnd is End - 1,
    slice(0,NewEnd,T,Return),
    !.
slice(0,0,[H|_],[H]):-!.
slice(Start,End,[_|T],Return):-
    NewStart is Start - 1,
    slice(NewStart,End,T,Return).


